src/Workspaces/Core/Portable/Rename/Renamer.RenameSymbolDocumentAction.cs

```diff
@@ -0,0 +1,120 @@
+                    return solution;
+                }
+
+                var document = solution.GetRequiredDocument(DocumentId);
+                var matchingTypeDeclaration = await GetMatchingTypeDeclarationAsync(document, _analysis.Symbol.Name, cancellationToken).ConfigureAwait(false);
+
+                if (matchingTypeDeclaration is object)
+                {
+                    var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                    var symbol = semanticModel.GetDeclaredSymbol(matchingTypeDeclaration, cancellationToken);
+
+                    solution = await RenameSymbolAsync(solution, symbol, _analysis.NewSymbolName, OptionSet, cancellationToken).ConfigureAwait(false);
+                }
+
+                return solution;
+            }
+
+            private static async Task<SyntaxNode> GetMatchingTypeDeclarationAsync(Document document, string name, CancellationToken cancellationToken)
+            {
+                var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+                var typeDeclarations = syntaxRoot.DescendantNodesAndSelf().Where(syntaxFacts.IsTypeDeclaration);
+                return typeDeclarations.FirstOrDefault(d => syntaxFacts.GetDisplayName(d, DisplayNameOptions.None).Equals(name, StringComparison.OrdinalIgnoreCase));
+            }
+
+            public static async Task<RenameSymbolDocumentAction> CreateAsync(Document document, string newName, OptionSet optionSet, CancellationToken cancellationToken)
+            {
+                var analysis = await RenameSymbolDocumentActionAnalysis.CreateAsync(document, newName, optionSet, cancellationToken).ConfigureAwait(false);
+
+                // TODO: Detect naming conflicts ahead of time
+                return new RenameSymbolDocumentAction(document, analysis, optionSet, ImmutableArray<string>.Empty);
+            }
+
+            private readonly struct RenameSymbolDocumentActionAnalysis
+            {
+                public string OriginalDocumentName { get; }
+                public string NewDocumentName { get; }
+                public string NewSymbolName { get; }
+                public ISymbol? Symbol { get; }
+                private RenameSymbolDocumentActionAnalysis(
+                    Document document,
+                    string newName,
+                    ISymbol? symbol = null)
+                {
+                    OriginalDocumentName = document.Name;
+                    NewDocumentName = newName;
+                    NewSymbolName = Path.GetFileNameWithoutExtension(newName);
+                    Symbol = symbol;
+                }
+
+                public static async Task<RenameSymbolDocumentActionAnalysis> CreateAsync(Document document, string newName, OptionSet optionSet, CancellationToken cancellationToken)
```

Comment:
```suggestion
                public static async Task<RenameSymbolDocumentActionAnalysis?> TryCreateAsync(Document document, string newName, OptionSet optionSet, CancellationToken cancellationToken)
```

See other comment: wondering if the "we couldn't find the symbol" is better just expressing this as "we don't have an analysis" since then you can get rid of the Symbol property and simplify other stuff too.